src/boot/bootctl-status.c

```diff
@@ -503,6 +504,224 @@ int verb_status(int argc, char *argv[], void *userdata) {
         return r;
 }
 
+static int ref_file(Hashmap *known_files, const char *fn, int increment) {
+        char *k = NULL;
+        int n;
+
+        assert(known_files);
+
+        /* just gracefully ignore this. This way the caller doesn't
+           have to verify whether the bootloader entry is relevant */
+        if (!fn)
+                return 0;
+
+        n = PTR_TO_INT(hashmap_get2(known_files, fn, (void**)&k));
+        n += increment;
+        if (increment >= 0 && !k) {
+                int r = hashmap_put(known_files, strdup(fn), INT_TO_PTR(n));
+                if (r < 0)
+                        return r;
+        } else if (increment < 0 && n <= 0) {
+                (void) hashmap_remove(known_files, fn);
+                free(k);
+        } else {
+                int r = hashmap_update(known_files, fn, INT_TO_PTR(n));
+                if (r < 0)
+                        return r;
+        }
+
+        return n >= 0 ? n : 0;
+}
+
+static void deref_unlink_file(Hashmap *known_files, const char *fn, const char *root)
+{
+        _cleanup_free_ char *path = NULL;
+        int r;
+
+        /* just gracefully ignore this. This way the caller doesn't
+           have to verify whether the bootloader entry is relevant */
+        if (!fn || !root)
+                return;
+
+        r = ref_file(known_files, fn, -1);
+        if (r == 0) {
+                if (arg_dry)
+                        r = chase_symlinks_and_access(fn, root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, F_OK, &path, NULL);
+                else
+                        r = chase_symlinks_and_unlink(fn, root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, 0, &path);
+                if (r < 0) {
+                        log_error_errno(errno, "Failed to remove \"%s\": %m", path);
+                        return;
+                }
+                log_info("%s %s", arg_dry?"Would remove":"Removed", path);
+        }
+}
+
+static int count_known_files(const BootConfig *config, const char* root, Hashmap **ret_known_files) {
+        _cleanup_(hashmap_free_free_keyp) Hashmap *known_files = NULL;
+        int r = 0;
+        assert(config);
+        assert(ret_known_files);
+
+        known_files = hashmap_new(&path_hash_ops);
+        if (!known_files)
+                return log_oom();
```

Comment:
should probably just be `return -ENOMEM`.

in systemd we want to avoid duplicate (or missing) logging about errors, hence we have the rule that functions either log about all errors or about none (in which case the error is typically propagated to the caller, which then can decide to log). Thus, for each function figure out if it is a logging one, or if it should be a non-logging one. Typically, "main-function-like" functions should log, and "library-like" functions should not log on their own. Distinction is blurry, huge grey area, hence make up your mind, and stick to it throughout the function.

(debug logging, i.e. LOG_DEBUG-level logging is always ok btw, it does not count as logging in this contetx).

so, count_known_files() currently doesn't log about any error (but this oom check here), and neither do any of the functions called here (i.e. ref_file). hence i'd say: make this a non-logging function, so that the caller logs. Thus: do `return -ENOMEM` here, not `log_oom()`.